src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/HandlerOrchestrator.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClientX.Handlers.TransportCreation;
+
+namespace Microsoft.Data.SqlClientX.Handlers
+{
+    internal class HandlerOrchestrator
+    {
+        /// <summary>
+        /// Class responsible to create the chain of handlers, initiate chain, retain handler-context, maintain 
+        /// routing count and history of all handler contexts for all handlers in the chain
+        /// </summary>
+        private  IHandler<HandlerRequest> _initialHandler;
+        private ConnectionHandlerContext _initialConnectionHandlerContext;
+        private InternalConnectionContext _internalConnectionContext;
+        private int _routingCount;
+        private List<ConnectionHandlerContext> _routingHistory;
+
+        public HandlerOrchestrator(IHandler<HandlerRequest> initialHandler, ConnectionHandlerContext initialConnectionHandlerContext, InternalConnectionContext internalConnectionContext, List<ConnectionHandlerContext> history)
+        {
+            _initialHandler = initialHandler;
+            _initialConnectionHandlerContext = initialConnectionHandlerContext;
+            _internalConnectionContext = internalConnectionContext;
+            _routingHistory = history;
+            _routingCount = 0;
+        }
+
+        /// <summary>
+        /// Intiates the pre-defined chain of handlers
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        /// <exception cref="HandlerNotFoundException"></exception>
+        public async Task ProcessRequestAsync(bool isAsync, CancellationToken ct) 
```

Comment:
```suggestion
        public async ValueTask ProcessRequestAsync(bool isAsync, CancellationToken ct) 
```

Nit: switched to ValueTask to avoid a heap allocation when `isAsync` is false.

Reply:
Good point. Updated.